renderdoc/driver/gl/wrappers/gl_texture_funcs.cpp

```diff
@@ -1372,22 +1372,60 @@ void WrappedOpenGL::glCopyImageSubData(GLuint srcName, GLenum srcTarget, GLint s
       {
         TextureData &dstData = m_Textures[dstrecord->GetResourceID()];
 
-        if(srcX == 0 || srcY == 0 || srcZ == 0 || dstX == 0 || dstY == 0 || dstZ == 0)
-        {
-          // we only support whole copies - sub-copies will not work correctly.
-          RDCASSERT(srcWidth == RDCMAX(1, srcData.width >> srcLevel));
-          RDCASSERT(srcHeight == RDCMAX(1, srcData.height >> srcLevel));
-          RDCASSERT(srcDepth == RDCMAX(1, srcData.depth >> srcLevel));
+        GLsizei srcLevelWidth = RDCMAX(1, srcData.width >> srcLevel);
+        GLsizei srcLevelHeight = RDCMAX(1, srcData.height >> srcLevel);
+        GLsizei srcLevelDepth = RDCMAX(1, srcData.depth >> srcLevel);
 
-          RDCASSERT(srcWidth == RDCMAX(1, dstData.width >> dstLevel));
-          RDCASSERT(srcHeight == RDCMAX(1, dstData.height >> dstLevel));
-          RDCASSERT(srcDepth == RDCMAX(1, dstData.depth >> dstLevel));
+        GLsizei dstLevelWidth = RDCMAX(1, dstData.width >> dstLevel);
+        GLsizei dstLevelHeight = RDCMAX(1, dstData.height >> dstLevel);
+        GLsizei dstLevelDepth = RDCMAX(1, dstData.depth >> dstLevel);
 
+        if(srcX == 0 && srcY == 0 && srcZ == 0 && dstX == 0 && dstY == 0 && dstZ == 0 &&
+           srcLevelWidth == dstLevelWidth && srcLevelHeight == dstLevelHeight &&
+           srcLevelDepth == dstLevelDepth)
+        {
           dstData.compressedData[dstLevel] = srcData.compressedData[srcLevel];
         }
         else
         {
-          RDCWARN("glCopyImageSubData doesn't support offsetted copies of compressed data");
+          size_t srcImageSize = GetCompressedByteSize(srcLevelWidth, srcLevelHeight, srcLevelDepth,
+                                                      srcData.internalFormat);
+          size_t dstImageSize = GetCompressedByteSize(dstLevelWidth, dstLevelHeight, dstLevelDepth,
+                                                      dstData.internalFormat);
+          std::tuple<size_t, size_t> srcBlockSize = GetCompressedBlockSize(srcData.internalFormat);
+          std::tuple<size_t, size_t> dstBlockSize = GetCompressedBlockSize(dstData.internalFormat);
+
+          RDCASSERT(srcWidth % std::get<0>(srcBlockSize) == 0);
+          RDCASSERT(srcWidth % std::get<0>(dstBlockSize) == 0);
+          RDCASSERT(srcHeight % std::get<1>(srcBlockSize) == 0);
+          RDCASSERT(srcHeight % std::get<1>(dstBlockSize) == 0);
```

Comment:
btw `glCopyImageSubData` also allows copying between uncompressed and compressed textures, as long as the uncompressed texel size in bytes and compressed block size in bytes are the same. It's fine if you don't want to support this (after all that was the reason that sub-image copies was never implemented before), but I'd like you to add a check for this case and output an explicit warning.

Reply:
This seems a typo in fact. It should be like `RDCASSERT(dstWidth % std::get<0>(dstBlockSize) == 0);`
In fact I'm working with compression ETC2 on GPU, which relies on copying between uncompressed and compressed textures. And this is the reason why I worked this patch out.

Reply:
There is no `dstWidth` though at the moment, the function only takes a single width/height because it doesn't do any scaling. In the case of copying between compressed and uncompressed the destination width/height either gets divided by or multiplied by the block size depending on the order to get an implicit `dstWidth` / `dstHeight`.

If you do intend to support copies between compressed and uncompressed then that will require more work. For example the current code checks only that the *source* texture is compressed and goes into this path. If the destination texture is compressed and the source is not, this code won't execute at all. Also uncompressed textures do not have entries in `compressedData` - since they are not compressed! So you will need to do a GPU readback of the uncompressed source texture data temporarily to be able to then copy into a compressed texture.